src/v/cluster/tm_stm.cc

```diff
@@ -0,0 +1,385 @@
+        _tx_table.emplace(entry.id, entry);
+        _tx_locks.emplace(entry.id, ss::make_lw_shared<mutex>());
+        _end_locks.emplace(entry.id, ss::make_lw_shared<mutex>());
+    }
+    _last_snapshot_offset = data.offset;
+    _insync_offset = data.offset;
+}
+
+stm_snapshot tm_stm::take_snapshot() {
+    tm_snapshot tm_ss;
+    tm_ss.offset = _insync_offset;
+    for (auto& entry : _tx_table) {
+        tm_ss.transactions.push_back(entry.second);
+    }
+
+    iobuf tm_ss_buf;
+    reflection::adl<tm_snapshot>{}.to(tm_ss_buf, tm_ss);
+
+    stm_snapshot_header header;
+    header.version = supported_version;
+    header.snapshot_size = tm_ss_buf.size_bytes();
+
+    stm_snapshot stm_ss;
+    stm_ss.header = header;
+    stm_ss.offset = _insync_offset;
+    stm_ss.data = std::move(tm_ss_buf);
+    return stm_ss;
+}
+
+ss::future<> tm_stm::apply(model::record_batch b) {
+    const auto& hdr = b.header();
+    _insync_offset = b.last_offset();
+
+    if (hdr.type != tm_update_batch_type) {
+        return ss::now();
+    }
+
+    vassert(
+      b.record_count() == 1,
+      "We expect single command in a batch of tm_update_batch_type");
+    auto r = b.copy_records();
+    auto& record = *r.begin();
+    auto rk = reflection::adl<uint8_t>{}.from(record.release_key());
+
+    vassert(
+      rk == tx_updated_cmd::record_key, "Unknown tm update command: {}", rk);
+    tx_updated_cmd cmd = reflection::adl<tx_updated_cmd>{}.from(
+      record.release_value());
+
+    auto ptx = _tx_table.find(cmd.tx.id);
+    if (ptx == _tx_table.end()) {
+        if (cmd.prev_etag != cmd.tx.etag) {
+            vlog(
+              clusterlog.error,
+              "Inconsistent tm log. First command should depend on itself but "
+              "tx.id={}, etag={} it reffers to {}",
+              cmd.tx.id,
+              cmd.tx.etag,
+              cmd.prev_etag);
+            auto recovery_policy
+              = config::shard_local_cfg().tm_violation_recovery.value();
+            if (recovery_policy == model::violation_recovery_policy::crash) {
+                vassert(
+                  false, "Crushing to prevent potential consistency violation");
+            } else if (
+              recovery_policy == model::violation_recovery_policy::hang) {
+                vassert(
+                  false, "Crushing to prevent potential consistency violation");
+            } else {
+                vassert(
+                  recovery_policy
+                    == model::violation_recovery_policy::best_effort,
+                  "Unknown recovery policy {}",
+                  recovery_policy);
+                vlog(
+                  clusterlog.error,
+                  "Recovering by blindly applying tx.id={}, etag={}",
+                  cmd.tx.id,
+                  cmd.tx.etag);
+            }
+        }
+        _tx_table.emplace(cmd.tx.id, cmd.tx);
+        _end_locks.try_emplace(cmd.tx.id, ss::make_lw_shared<mutex>());
```

Comment:
why use `try_emplace` here and not in the previous statement?

Reply:
With previous statement I know for sure that tx_table doesn't have tx.id

Reply:
I don't really understand. These two statements have the exact same effect, so why is one chosen over the other?

Reply:
I thought emplace throws an exception if key is already there but I was wrong